CORE-7338 license: add `enterprise_license_expiry_sec` metric

pgellert opened the pull request:
This introduces a new metric for monitoring the expiry time of the enterprise license.

#### cluster_features_enterprise_license_expiry_sec
* Description: “Number of seconds remaining until the Enterprise License expires”
* Endpoints: both public and internal metrics
* Labels: none

Fixes https://redpandadata.atlassian.net/browse/CORE-7338

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v24.2.x
- [x] v24.1.x
- [x] v23.3.x

## Release Notes
### Features

* A new metric (*cluster_features_enterprise_license_expiry_sec*) is added for easier monitoring of the enterprise license's expiry time.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

oleiman reviewed (commented):
seems straightforward. couple of structural suggestions, at your discretion

pgellert commented on src/v/features/feature_table.cc:
```diff
@@ -697,6 +703,40 @@ void feature_table::assert_compatible_version(bool override) {
 
+feature_table::probe::probe(const feature_table& parent)
+  : _parent(parent) {
+    if (ss::this_shard_id() != 0) {
+        return;
+    }
+
+    if (!config::shard_local_cfg().disable_metrics()) {
+        setup_metrics(_metrics);
+    }
+
+    if (!config::shard_local_cfg().disable_public_metrics()) {
+        setup_metrics(_public_metrics);
+    }
```
Yeah, that's a fair point. There's no `feature_table::start` method at the moment, and similarly there are no `start` methods on any controller-log-backed `_table.cc`'s. I think that's because they are `ss::sharded<>` and they are expected to be constructed and started at the same time with `ss::sharded::start()`. So what I've done now is I've added a `setup_metrics` method that gets called inside the constructor of `feature_table` right after the probe is created to make it explicit that we start the probe immediately.

oleiman replied:
>expected to be constructed and started at the same time with `ss::sharded::start()`

Not necessarily, you'll see this pattern throughout `application.cc`

https://github.com/redpanda-data/redpanda/blob/dev/src/v/redpanda/application.cc#L2475-L2478

But yeah, it doesn't seem like this probe should need an async lifetime since the lifetime of the `sharded<feature_table>` tracks with the whole program...the test failures[1] are real and maybe a result of feature_table being con/destructed manually across various tests and fixtures. 

[1] [build](https://buildkite.com/redpanda/redpanda/builds/54738#01920aec-ccf8-43d5-828d-1b636d7d9d3b) - these do go away if I set `_probe = nullptr` in `feature_table::feature_table(...)`

pgellert replied:
Those are not `_table`'s though 😉 but yeah fair point, there are certainly examples of decoupled sharded service construction and start. I'm wondering what seastar-people would think of that.

Thanks for the pointer on the build, the cause turned out to be multiple `feature_table`'s being constructed at the same time.

pgellert commented:
Force-push:
* [First](https://github.com/redpanda-data/redpanda/compare/0c560e5f527c33c6293c4c89b945009651cad0eb..c29dc0c869a79b920e81831119fadf4740a47414): rebase to dev
* [Second](https://github.com/redpanda-data/redpanda/compare/c29dc0c869a79b920e81831119fadf4740a47414..1d76d0af0834778902509233dfe6d4df0173e754): Fixed the linking error in the CI build (missing dependency on security), addressed code review feedback

pgellert commented on src/v/security/license.cc:
```diff
@@ -179,10 +184,17 @@ license make_license(const ss::sstring& raw_license) {
+    auto now_sec = std::chrono::duration_cast<std::chrono::seconds>(
+      now.time_since_epoch());
+    return std::max(0s, expiry - now_sec);
+}
```
Fixed. I think we're good to use `system_clock` regardless of the leap seconds issue, but yeah, it's definitely good to know.

oleiman replied:
oh yeah, I was just referring to the fact that unix epoch is always the basis for `system_clock::time_point`. as i understand it, this was not in the standard pre-C++20.

pgellert commented on src/v/features/BUILD:
```diff
@@ -21,6 +21,7 @@ redpanda_cc_library(
         "//src/v/cluster:features",
         "//src/v/cluster:version",
         "//src/v/config",
+        "//src/v/metrics",
```
I'm not 100% sure if this could be `implementation_deps` only. I'll need to double-check.

pgellert replied:
This point no longer applies. I've made it an `implementation_deps` now that the probe is hidden inside the implementation file.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/54738#01920b30-447c-414f-a20b-3d774ec8f978
<!--0ac9bc58-7f53-4505-af5c-b1c21e6fc53c-->

oleiman commented:
Threw @BenPope on the review because he's helped me with metrics footguns before...not sure exactly what the problem is here, but my knee-jerk guess is lifetime issues.

kbatuigas commented on src/v/features/feature_table.cc:
```diff
@@ -697,6 +706,49 @@ void feature_table::assert_compatible_version(bool override) {
+          "enterprise_license_expiry_sec",
+          [this]() { return calculate_expiry_metric(_parent.get_license()); },
+          sm::description(
+            "Number of seconds remaining until the Enterprise License expires"))
```
```suggestion
            "Number of seconds remaining until the Enterprise license expires"))
```

pgellert replied:
Thanks, I've updated this now

kbatuigas reviewed (commented):
Just a suggestion (we won't need to capitalize "license" in the docs), lgtm otherwise

BenPope commented on src/v/features/feature_table.cc:
```diff
@@ -697,6 +706,49 @@ void feature_table::assert_compatible_version(bool override) {
     }
 }
 
+feature_table::probe::probe(const feature_table& parent)
```
`feature_table` can be passed through `setup_metrics`.

pgellert replied:
I've left it to be passed in the constructor because then I can hold it as a reference instead of a pointer inside the probe which means that I don't have to deal with a `nullptr`. tldr: it's simpler to pass it in the constructor and `setup_metrics` gets called right after the constructor so there's not much of a difference here

BenPope replied:
I thinking to capture it directly into the lambda:
```c++
             [&ft]() {
                  return calculate_expiry_metric(ft.get_license());
              },
```
But it's a minor point.

BenPope commented on src/v/features/feature_table.cc:
```diff
@@ -697,6 +706,49 @@ void feature_table::assert_compatible_version(bool override) {
+long long feature_table::probe::calculate_expiry_metric(
+  const std::optional<security::license>& license,
+  security::license::clock::time_point now) {
+    if (!license) {
+        return -1;
+    }
```
discussion:

If there is no license, it may be possible to not set up this metric (or even the probe), but then setting up the probe or metric must be attached to `set_license` / `revoke_license / `feature_table_snapshot::apply`.

For the latter, it may be worth:
```diff
diff --git a/src/v/features/feature_table_snapshot.cc b/src/v/features/feature_table_snapshot.cc
index 6e1cfb406e..b5477a88c5 100644
--- a/src/v/features/feature_table_snapshot.cc
+++ b/src/v/features/feature_table_snapshot.cc
@@ -42,7 +42,11 @@ void feature_table_snapshot::apply(feature_table& ft) const {
       applied_offset);
 
     ft.set_active_version(version);
-    ft._license = license;
+    if (license) {
+        ft.set_license(*license);
+    } else {
+        ft.revoke_license();
+    }
 ```

pgellert replied:
Yeah, that's a good point to discuss but I've kept this as is for now to keep things simpler. I kinda like having the -1 explicitly present to signal that the license it unset (as opposed to just the metric missing because of network issues, Prometheus scraper being down, etc.). But I don't feel strongly either way which makes me prefer simplicity.

pgellert commented on src/v/features/feature_table.h:
```diff
@@ -655,6 +683,7 @@ class feature_table {
 
     ss::gate _gate;
     ss::abort_source _as;
+    std::unique_ptr<probe> _probe;
```
Thanks for the detailed explanation, very useful. I've changed this now to have the probe be forward declared and its implementation hidden inside the implementation file.

pgellert commented on src/v/security/license.cc:
```diff
@@ -179,10 +184,14 @@ license make_license(const ss::sstring& raw_license) {
 
 bool license::is_expired() const noexcept {
     const auto now = std::chrono::duration_cast<std::chrono::seconds>(
-      std::chrono::system_clock::now().time_since_epoch());
+      clock::now().time_since_epoch());
     return now > expiry;
```
Fixed

pgellert commented:
[Force-push](https://github.com/redpanda-data/redpanda/compare/1d76d0af0834778902509233dfe6d4df0173e754..f2fc540ae2c42c95dd589ba8ca8ac6e67fcb4c30):
* code review feedback (mostly hiding the probe in the implementation file plus some minor things)
* added a bazel build for features/test while I was there
* fixed the failing test -- some storage fixture tests create multiple `feature_table` instances at the same time (specifically `disk_log_builder` creates one under the hood) which causes metric double-registration issues, so to solve this I followed the existing pattern and disabled metrics

BenPope commented on src/v/features/feature_table.cc:
```diff
@@ -232,9 +284,15 @@ feature_table::feature_table() {
             }
         }
     }
+
```
declaring the destructor made this true, but I wonder if it's worth being explicit:
```c++
    static_assert(
      !std::is_move_constructible_v<feature_table>
        && !std::is_move_assignable_v<feature_table>
        && !std::is_copy_constructible_v<feature_table>
        && !std::is_copy_assignable_v<feature_table>,
      "The probe captures a reference to this");
```

pgellert replied:
Yeah I like it, I've adopted this now

BenPope commented on tests/rptest/tests/rpk_cluster_test.py:
```diff
@@ -209,6 +209,25 @@ def test_cluster_down(self):
+
+        internal_val = get_metrics_value(MetricsEndpoint.METRICS)
+        public_val = get_metrics_value(MetricsEndpoint.PUBLIC_METRICS)
+        assert internal_val == public_val, f"Mismatch: {internal_val} != {public_val}"
```
I guess it's possible to get unlucky and have the scrapes cross over a 1 second boundary.

pgellert replied:
That should be fine though because it's called from a `wait_until` with `retry_on_exc=True` so in that case we would just retry.

oleiman reviewed (approved):
lgtm

pgellert commented on src/v/features/feature_table.h:
```diff
@@ -497,6 +498,7 @@ class feature_table {
     static cluster::cluster_version get_earliest_logical_version();
 
     feature_table();
```
Done

pgellert commented on src/v/features/feature_table.cc:
```diff
@@ -697,6 +755,18 @@ void feature_table::assert_compatible_version(bool override) {
+
+    auto rem = license->expiration() - now;
+    auto rem_capped = std::max(rem.zero(), rem);
+    return std::chrono::duration_cast<std::chrono::seconds>(rem_capped).count();
```
Done

pgellert commented:
[Force-push](https://github.com/redpanda-data/redpanda/compare/f2fc540ae2c42c95dd589ba8ca8ac6e67fcb4c30..f630e9317d64ebd5561a1d3072e09140bb493ea0): adopt remaining minor improvement suggestions

oleiman reviewed (approved):
lgtm!

pgellert commented:
Force-push:
* [Rebase to dev](https://github.com/redpanda-data/redpanda/compare/f630e9317d64ebd5561a1d3072e09140bb493ea0..89ccd7391eca9cf82d02f43cff29fe279e6ec0e4)
* [Fix bazel build error due to conflict with recent changes on dev](https://github.com/redpanda-data/redpanda/compare/89ccd7391eca9cf82d02f43cff29fe279e6ec0e4..baece2ccd6c9986a85c1af108dbb5b8387e85f1f)

BenPope commented on src/v/features/feature_table.cc:
```diff
@@ -200,6 +206,60 @@ static std::array test_extra_schema{
+          {
+            sm::make_gauge(
+              "enterprise_license_expiry_sec",
+              [&ft = _parent]() {
```
My idea was so that you didn't need the `_parent` member.

pgellert replied:
Right, now I get what you meant there, that makes more sense. I will leave this as is now since the behaviour is pretty much the same.

vbotbuildovich commented:
/backport v24.2.x

vbotbuildovich commented:
/backport v24.1.x

vbotbuildovich commented:
/backport v23.3.x

vbotbuildovich commented:
Failed to create a backport PR to v24.1.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-23367-v24.1.x-169 remotes/upstream/v24.1.x
git cherry-pick -x 78de1c702ef22c1becb876913674b9c5fa7de31b 1c8f66b467b0b5e84de6e6c424fbe65e825f5e98 edb5f48493aaec25ebcb10f35eb26b3ad5ae1a10 c44d9d8f53222dc4c50063cafc8ff020dc4e586d baece2ccd6c9986a85c1af108dbb5b8387e85f1f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/11016952213).

vbotbuildovich commented:
Failed to create a backport PR to v24.2.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-23367-v24.2.x-227 remotes/upstream/v24.2.x
git cherry-pick -x 78de1c702ef22c1becb876913674b9c5fa7de31b 1c8f66b467b0b5e84de6e6c424fbe65e825f5e98 edb5f48493aaec25ebcb10f35eb26b3ad5ae1a10 c44d9d8f53222dc4c50063cafc8ff020dc4e586d baece2ccd6c9986a85c1af108dbb5b8387e85f1f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/11016952202).

vbotbuildovich commented:
Failed to create a backport PR to v23.3.x branch. I tried:

```
git remote add upstream https://github.com/redpanda-data/redpanda.git
git fetch --all
git checkout -b backport-pr-23367-v23.3.x-530 remotes/upstream/v23.3.x
git cherry-pick -x 78de1c702ef22c1becb876913674b9c5fa7de31b 1c8f66b467b0b5e84de6e6c424fbe65e825f5e98 edb5f48493aaec25ebcb10f35eb26b3ad5ae1a10 c44d9d8f53222dc4c50063cafc8ff020dc4e586d baece2ccd6c9986a85c1af108dbb5b8387e85f1f
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/11016952984).
